Refactored JS tiny world with OOP style

SerhiiYakovenko opened the pull request:
### JS post OOP exercise

[Demo](https://serhiiyakovenko.github.io/a-tiny-JS-world/)
[Code base](https://github.com/SerhiiYakovenko/a-tiny-JS-world/tree/syakovenko-js-world-oop)

The code is submitted in a dedicated feature branch.
Please, review.

SerhiiYakovenko commented:
I was trying to use a simpler view of the same program:
```
function Inhabitant(species, name, gender, legs, hands, saying, ...friends) {
    this.species = species;
    this.name = name;
    this.gender = gender;
    this.legs = legs;
    this.hands = hands;
    this.saying = saying;
    this.friends = friends;
};
```
With defining friends in the main class, but on use with 
```
for my Best friends: ${this.friends[0].map(friend => friend.name).join(', ')}`
```

I got the error `Cannot access before initialization`.
So in case there I can do anything by removing the additional method `addFriends` I would like to know how I can make my code even more efficient.
Thanks!

MaxKalashnyk reviewed (changes requested):
Please, create a separate repository for this task. No need to change the file from the `tiny-js-world` task.

SerhiiYakovenko commented:
@MaxKalashnyk 
According to this page
https://github.com/kottans/frontend/blob/master/tasks/js-post-oop.md

```
Improve your tiny JS world model you forked from here. This time use your favourite JS OOP methods to keep your code DRY and SOLID.

You don't need forking this time, work with the repo you've already got. Git keeps history for you.
```
Also, there was a conversation in telegram:
https://t.me/c/1382428271/28703

Git history for both of these tasks should be ok.
Can you please confirm it?
Thanks!

MaxKalashnyk commented:
> @MaxKalashnyk
> According to this page
> https://github.com/kottans/frontend/blob/master/tasks/js-post-oop.md
> 
> ```
> Improve your tiny JS world model you forked from here. This time use your favourite JS OOP methods to keep your code DRY and SOLID.
> 
> You don't need forking this time, work with the repo you've already got. Git keeps history for you.
> ```
> 
> Also, there was a conversation in telegram:
> https://t.me/c/1382428271/28703
> 
> Git history for both of these tasks should be ok.
> It’s not an issue at all to move it to separate folder, but would like to do it according requirements.
> Can you please confirm it?
> Thanks!

Yes, you're right. My fault. I'll remove the label.

MaxKalashnyk reviewed (changes requested):
I clarified the information about the files structure.
So you should create a new repository for this task but you can work using your files from the `tiny-js-world` task. Sorry for misunderstandings.

SerhiiYakovenko commented:
Done 👍

OleksiyRudenko reviewed (changes requested):
@SerhiiYakovenko nice transition from literals to prototypical syntax.
OOP is not only about syntax though.
For example, following the OOP/SOLID principles the inhabitants wouldn't have the props they do not have by natural design. Imagine what a nightmare for your peer developer would be adding birds and then humans with zero wings.

I'd recommend you revisit [extra materials from the original project](https://github.com/OleksiyRudenko/a-tiny-JS-world#leveling-your-skills-up), switch over to es6 class syntax and re-build the code so that it is scalable and maintanable.

SerhiiYakovenko commented:
Thanks, @OleksiyRudenko, now I see a difference. 
Your JS-ES snippet was super useful.

SerhiiYakovenko commented:
Hey, guys @MaxKalashnyk @OleksiyRudenko, I already asked twice for a review in the FE questionarium, maybe you missed it.
PR is ready for re-review for 23 days :)
Can you please take a look?

OleksiyRudenko commented:
@SerhiiYakovenko please check the latest AMA session video. It addresses the issue with `hands` prop specifically. Do not miss other typical issues covered during the session

SerhiiYakovenko commented:
Done, removed `hands` from default `Inhabitant` class and added it directly to `Human`
Removed `[0]` from `friends#map`.
Thanks for AMA @OleksiyRudenko !
Re-review pls

OleksiyRudenko commented on submissions/SerhiiYakovenko/tiny-js-oop/index.js:
```diff
@@ -0,0 +1,92 @@
+     this.friends = [];
+   }
+ 
+   addFriends(friends) {
+     this.friends = [...friends];
+   }
```
As a developer I would expect the the following
```
kot.addFriends([boi, pes]);
kot.addFriends([woman]);
```
would result `kot` to have 3 freinds as method name suggests.
This won't happen.

OleksiyRudenko commented on submissions/SerhiiYakovenko/tiny-js-oop/index.js:
```diff
@@ -0,0 +1,92 @@
+ kot.addFriends([boi, pes]);
+ 
+ [boi, mama, papa, pes, kot].forEach((person) => print(person.createOutput()));
+ 
```
Looks like you still need to tune up your IDE.
![image](https://user-images.githubusercontent.com/8686631/108955707-4e31bb80-7677-11eb-873b-149090a3f05e.png)

This may cause issues when you collaborate with others.

Consult git and your IDE documentation and make sure that newlines are `LF` and not `CR` or `CRLF`.

OleksiyRudenko reviewed (approved):
@SerhiiYakovenko approved with a minor reservation
